### ISSUE TITLE
Add links from advices to the advised elements

### ISSUE THREAD

**Issue by lisachenko** [goaop/idea-plugin#5]
It would be nice to have a links from advices to the concrete elements. Unfortunately, it's very expensive operation, because I need to take one single pointcut and then match it for all `PhpNamedElement`s such as `Method` and `Field`. 

There is a branch `feature/navigate-to-advised-elements` with working implementation, but some kind of optimization should be done, before using this for real projects.

**Comment by lisachenko** [comment]
Hi, Daniel! /cc @Haehnchen Sorry for the trouble!

Could I ask your for an expert advice? Previously, you pointed me into the right direction and this gave me the ability to finish the main part of my plugin! Now PhpStorm can analyze any pointcuts and shows links to aspect classes + their advices, it's so cool! Thank you very much! )

Now I want to provide line marker from advices to the advised elements and realized, that this task can't be done easily by taking the pointcut expression and matching it for all `PhpClassMemeber`s such as `Method` and `Field`, because it requires to perform checks for thousands of elements and naviagtion icon will appear only after several seconds.

Is it possible somehow in IDE to create a cache with matching elements in background? My primary idea was to take my index with pointcuts, then perform checks for each element in `PhpMethodIndex` and `PhpFieldIndex` indexes, then store result in a `HasMap<Pointcut, List<PhpNamedElement>>` and invalidate it when pointcut or matched file is changed.

Thank you in advance!

**Comment by Haehnchen** [comment]
i am using this for symfony services and a symfony like event subscriber use case:
- add a pre index filter to reduce files to scans: https://github.com/Haehnchen/idea-php-shopware-plugin/blob/master/src/de/espend/idea/shopware/index/EventConfigGoToIndex.java
- use a lazy target linemarker; so add it if there is a knowns object and load targets on icon click (supported by main LineMarker extension already)
- cache psi results, also long no file changes
  https://github.com/Haehnchen/idea-php-shopware-plugin/commit/024e5dc68048261a7b41936b515e4c9dfb64dee0#diff-885e0221b9031a39515e082d45ecacb9R164

**Comment by lisachenko** [comment]
Thank you very much! :+1: I need some time to think about your solution and check if it can be used for me. Will answer here about results

**Comment by lisachenko** [comment]
Hmm, your pre index filter knows which element to look for, but for my case I don't know exactly the names of methods to search. Is it possible to use an index to create another index, or this will produce an exception? 

Lazy target linemarker - looks promising, is it ok to check 20-30k elements in runtime for lazy linemarker?

Your last link is very good, it has many common with my hooks ) But it's unclear how to invalidate this cache when one of the following changed: aspect class, which contains a pointcut with advice, concrete class with methods. I'm not familiar wit caching system, but looks like there are special events which I can handle to invalidate only part of that cache.
